Logan/#6/add segment control in main view

sloDevz opened the pull request:
### Segment Control
- Segment control UI 구현 및 Segment control의 selection에 따라 List의 정렬순서가 달라지도록 기능 구현

---
### Category Button
- Category Button UI 구현
- Button의 세부적인 기능(Drop down)은 이후 구현예정
---
### TagFilterButton
- TagFilterButton UI 구현
- 세부적인 기능(BottomSheet 및 선택한 Tag의 컨텐츠만 sorting)은 이후 구현예정
---
### ReleaseDate games
- 발매일 순 List일때 출시된지 3주 미만인 Game은 new뱃지 나타나도록 구현

sloDevz committed:
feat: #20 segmentControl 기본구현
- segmentItems 구성을 위한 ListType CaseIterable 채택

sloDevz committed:
feat: #6 segmentControl 변경사항 및 UINavigationController 추가
- SegmentControl의 SelectionColor 변경
- Hex Color도 받을 수 있도록 UIColor extension 추가
- URLRsponse 파일명에 '+' 추가
- initialView를 UINavigationController로 변경

sloDevz committed:
feat: 카테고리 변경버튼 기본UI 구현 및 Segment Control 위치 변경
- #7의 카테고리 변경버튼 기본적인UI 구현
- Segment Control의 위치를 NavigationBar로 변경
- GameListCollectionViewController에 쓰이는 Constants 정의

sloDevz committed:
chore: filterButton이미지 Assets에 등록

sloDevz committed:
feat: tagFilterButton UI추가

sloDevz committed:
feat: segmentControl 에 따라 GameList 달라지도록 구현

sloDevz committed:
chore: 일부 타입 및 변수명을 조금더 직관적으로 변경
- ListType을 ListSortingType으로 변경
- rankedBedge를 bedge로 변경

sloDevz committed:
chore: setRankedBedgeStyle -> setBedgeStyle 메서드명 변경

sloDevz committed:
feat: #6 최신게임순 new뱃지 구현 및 오타 수정
- 최신 게임List에서 출시된지 3주 미만인 게임에 new 뱃지 나타나도록 구현
- NewBadge 이미지 Assets에 추가
- 코드내 typo 수정 bedge -> badge
- DateHandler : 날짜관련 로직을 전담하는 클래스 추가
- ListSortingType -> 전역 타입으로 변경

ohdair commented on GameScope/GameScope/Controllers/GameListCollectionViewController.swift:
```diff
@@ -34,14 +42,55 @@ final class GameListCollectionViewController: UIViewController {
+        return button
+    }()
+
+    private lazy var seperatorLineView: UIView = {
```
`lazy var`로 설정한 이유가 궁금해요

ohdair commented on GameScope/GameScope/Controllers/GameListCollectionViewController.swift:
```diff
@@ -34,14 +42,55 @@ final class GameListCollectionViewController: UIViewController {
+        segmentControl.backgroundColor = .white
+        segmentControl.selectedSegmentTintColor = .init(hex: "9FEBD9")
+
+        for index in .zero...segmentItems.count-1 {
+            segmentControl.setWidth(view.frame.width*0.4, forSegmentAt: index)
+        }
```
하나의 숫자에 값을 지정할 때 `.zero`를 쓰고 코드에서는 0으로 표기하는 게 보기 더 좋을 거 같아요.
`segmentItems.count-1`보다는 `(segmentItems.count - 1)` 곱하기 도 동일한 거 같아요
`width * 0.4` 더 보기 좋다 생각이 들어요.

ListSortingType 에서 Int를 채택한 뒤, 아래와 같이 쓰는 방법도 있을 거 같아요
```swift
for type in ListSortingType.allCases {
    segmentControl.setWidth(view.frame.width * 0.4, forSegmentAt: type.rawvalue)
}
```

ohdair commented on GameScope/GameScope/Controllers/GameListCollectionViewController.swift:
```diff
@@ -34,14 +42,55 @@ final class GameListCollectionViewController: UIViewController {
             GameListCollectionViewCell.self,
             forCellWithReuseIdentifier: GameListCollectionViewCell.identifier)
-
+        collectionView.clipsToBounds = false
```
clipsToBounds 에서 default 값은 false로 알고 있어요

ohdair commented on GameScope/GameScope/Controllers/GameListCollectionViewController.swift:
```diff
@@ -128,7 +220,7 @@ extension GameListCollectionViewController {
         snapShot.appendSections([.main])
         snapShot.appendItems(games)
-        dataSource.apply(snapShot)
+        dataSource.apply(snapShot,animatingDifferences: false )
```
잘못된 공백으로 스타일 가이드와는 맞지 않아요.

ohdair commented on GameScope/GameScope/Extension/UIColor+.swift:
```diff
@@ -0,0 +1,31 @@
+
+import UIKit
+
+extension UIColor {
+    convenience init?(hex: String) {
+        var hexString = hex.trimmingCharacters(in: .whitespacesAndNewlines).uppercased()
+
+        if hexString.hasPrefix("#") {
+            hexString.remove(at: hexString.startIndex)
+        }
+
+        guard hexString.count == 6 else {
+            return nil
+        }
+
+        var rgbValue: UInt64 = 0
+        Scanner(string: hexString).scanHexInt64(&rgbValue)
+
+        let red = CGFloat((rgbValue & 0xFF0000) >> 16) / 255.0
+        let green = CGFloat((rgbValue & 0x00FF00) >> 8) / 255.0
+        let blue = CGFloat(rgbValue & 0x0000FF) / 255.0
+
+        self.init(red: red, green: green, blue: blue, alpha: 1.0)
+    }
+}
```
메인 컬러를 설정에만 사용되는 거라면 Asset으로 등록해서 사용하는 게 더 좋을 거 같아요.
다른 방법으로는 `#colorLiteral(` 도 있어요

ohdair commented on GameScope/GameScope/Util/DateHandler.swift:
```diff
@@ -0,0 +1,53 @@
+
+    // MARK: - Public
+    func dayCount(from startDateString: String? = nil, to endDateString: String) -> Int? {
+
```
불필요한 개행인 거 같아요

ohdair commented on GameScope/GameScope/View/GameListCollectionViewCell.swift:
```diff
@@ -21,21 +21,22 @@ final class GameListCollectionViewCell: UICollectionViewCell {
-        static let rankBedgeInset: CGFloat = 6
+        static let badgeInset: CGFloat = 6
         static let thumbnailAndInfoContainerInset: CGFloat = 10
         static let TitleAndDescribsionInset: CGFloat = 4
```
`TitleAndDescribsionInset` 속성이 대문자로 시작해요

ohdair commented on GameScope/GameScope/Controllers/GameListCollectionViewController.swift:
```diff
@@ -68,14 +115,59 @@ final class GameListCollectionViewController: UIViewController {
+    @objc private func listTypeChanged() {
+        let selectedIndex = segmentControl.selectedSegmentIndex
+        switch selectedIndex {
+        case .zero:
```
GameListCollectionViewCell에서 사용했던 것과 같이 0으로 하는 게, 통일성에 맞을 거 같아요

ohdair commented on GameScope/GameScope/View/GameListCollectionViewCell.swift:
```diff
@@ -102,22 +103,42 @@ final class GameListCollectionViewCell: UICollectionViewCell {
+                badge.isHidden = true
+                return
+            }
+            badge.isHidden = false
```
default 값이 false로 불필요한 코드 같아요

ohdair commented on GameScope/GameScope/View/GameListCollectionViewCell.swift:
```diff
@@ -102,22 +103,42 @@ final class GameListCollectionViewCell: UICollectionViewCell {
-            return
+    private func setBadgeStyle(index: Int, releaseDate: String, listSortingType: ListSortingType) {
+        if listSortingType == .popularity {
+            guard index < 3 else {
+                badge.isHidden = true
+                return
+            }
```
default에서 동일한 작업을 하고 있어요

ohdair commented on GameScope/GameScope/View/GameListCollectionViewCell.swift:
```diff
@@ -102,22 +103,42 @@ final class GameListCollectionViewCell: UICollectionViewCell {
+            guard let days = DateHandler().dayCount(to: releaseDate) else { return }
+
+            if days > -21 {
+                badge.isHidden = false
```
default 값이 false로 불필요한 코드 같아요

ohdair commented on GameScope/GameScope/View/GameListCollectionViewCell.swift:
```diff
@@ -102,22 +103,42 @@ final class GameListCollectionViewCell: UICollectionViewCell {
+                badge.isHidden = false
+                badge.image = UIImage(named: Constants.newBadgeImageName)
+                return
+            }else {
```
공백이 필요한 거 같아요

ohdair commented on GameScope/GameScope/View/GameListCollectionViewCell.swift:
```diff
@@ -102,22 +103,42 @@ final class GameListCollectionViewCell: UICollectionViewCell {
+            if days > -21 {
+                badge.isHidden = false
+                badge.image = UIImage(named: Constants.newBadgeImageName)
+                return
+            }else {
+                badge.isHidden = true
+                return
```
return은 없어도 될 거 같아요

ohdair reviewed (commented):
이번 PR에서는 스타일 가이드와는 맞지 않은 코드가 많았던 거 같아요.
사소한 것처럼 보여도 PR 보내기 전 코드를 한 번 읽어보는 걸 부탁해요

sloDevz committed:
refactor: 코드리뷰 반영
- 일부 코드 가독성 개선 및 컨벤션 준수하기 위해 수정
- Color를 Asset등록하여 사용하도록 변경 -> 변경에 따른 UIColor extesion 코드 및 파일 제거
- 불필요한 개행 및 속성이름 typo 및 대소문자 수정

sloDevz committed:
refactor: badge 및 LatestGame 관련 로직 변경
- Badge.isHidden 값 초기화 시키는 로직 위치 변경
- 최신게임의 기준을 30일로 변경
- 최신게임 기준일을 표현하는 Constants 추가

ohdair commented on GameScope/GameScope/Controllers/GameListCollectionViewController.swift:
```diff
@@ -34,14 +42,54 @@ final class GameListCollectionViewController: UIViewController {
+        segmentControl.backgroundColor = .white
+        segmentControl.selectedSegmentTintColor = UIColor(named: "keyColor")
+
+        for index in 0...(segmentItems.count-1) {
```
숫기 기호의 경우에는 앞 뒤로 공백을 넣어주세요

ohdair commented on GameScope/GameScope/Controllers/GameListCollectionViewController.swift:
```diff
@@ -128,7 +219,7 @@ extension GameListCollectionViewController {
         snapShot.appendSections([.main])
         snapShot.appendItems(games)
-        dataSource.apply(snapShot)
+        dataSource.apply(snapShot, animatingDifferences: false )
```
false 뒤에 공백이 있어요

ohdair commented on GameScope/GameScope/View/GameListCollectionViewCell.swift:
```diff
@@ -97,27 +99,43 @@ final class GameListCollectionViewCell: UICollectionViewCell {
-            rankBedge.isHidden = true
-            return
+    private func setBadgeStyle(index: Int, releaseDate: String, listSortingType: ListSortingType) {
+        if listSortingType == .popularity {
+            guard index < 3 else {
+                return
+            }
```
guard 문을 if문에 합치는 게 더 좋은 표현일 거 같아요

ohdair commented on GameScope/GameScope/View/GameListCollectionViewCell.swift:
```diff
@@ -97,27 +99,43 @@ final class GameListCollectionViewCell: UICollectionViewCell {
-        default:
-            rankBedge.isHidden = true
+
+        if listSortingType == .releaseDate {
+            guard let days = DateHandler().dayCount(to: releaseDate) else { return }
+
+            if days > -Constants.latestDayRange {
```
if 문 하나로 표현이 가능할 거 같아요. 지금으로는 depth가 더 깊어지는 것보다는 한 번에 표현이 좋을 거 같아요

ohdair reviewed (commented):
자주 사용되는 Key Color의 표현이 Asset으로 표현하니 더 깔끔하니 좋은 거 같아요

sloDevz committed:
refactor: 리뷰반영
- 불필요한 공백 및 컨벤션 정리
- setBadgeStyle 메서드 코드 정리

sloDevz committed:
chore: 공백수정
